### DISCUSSION TITLE
named.Required VS named.Optional

### DISCUSSION THREAD

**Discussion by dangerousvasil** [ydb-platform/ydb-go-sdk#1115] [Q&A] [answered]
https://ydb.tech/docs/ru/reference/ydb-sdk/example/go/
Подскажите по ГО СДК. вопрос такой
создал таблицу с
source_type Utf8, -- не опциональный
пытаюсь его сосканить в переменную

1) делаю source_type как строку 
и вызываю
получается указатель на строку
named.Required("source_type", &source_type),
получаю ошибку incorrect source types PRIMITIVE_TYPE_ID_UNSPECIFIED
"tx: non-retryable error occurred on attempt No.1 (idempotent=false): scan row failed: incorrect source types PRIMITIVE_TYPE_ID_UNSPECIFIED at github.com/ydb-platform/ydb-go-sdk/v3/internal/table/scanner.(*valueScanner).setString(scanner.go:753) at github.com/ydb-platform/ydb-go-sdk/v3/internal/table.do.func1(retry.go:54) at github.com/ydb-platform/ydb-go-sdk/v3/internal/table.do.retryBackoff.func2(retry.go:85) at github.com/ydb-platform/ydb-go-sdk/v3/retry.Retry(retry.go:314) at `github.com/ydb-platform/ydb-go-sdk/v3/internal/table.(*Client).Do(client.go:658)`"

2) делаю source_type как указатель на строку
получается указатель на указатель на строку
named.Optional("source_type", &source_type),
так работает 

вопрос?
что я неправильно понимаю?
вроде как Optional это для типов которые созданны как Option<Utf8>?

**Comment by asmyasnikov** [comment, marked as answer]
все хелперы принимаю ссылку на destination - чтобы туда записать данные из строки ydb
в частности 
named.Required - принимает ссылку на тип. Пример:
```
var s string
named.Required("a", &s)
```
named.Optional  - принимает ссылку на указатель на тип. 
Второе сделано ровно для того, чтобы на стороне пользователя различать NULL значение и default type value
Пример.
```
var s *string
named.Optional("a", &s)
```
если в s - nil, значит в соответствующей колонке строки ydb был NULL
если не nil - в s будет адрес реальных данных
